### PR TITLE
[Sprint 9 / Enhancement] Refactor Akushon

### PR DESCRIPTION
**Pull request by maroqijalil**
## Jira Link: 

https://ichiro-its.atlassian.net/browse/PI-113

## Description

Akushon has main node and action node, the main node, called Akushon node, has main functionality to handle request for running an action. The action node has functionality to handle main algorithm for running an action and make request for getting joints value from Tachimawari package.

## Type of Change

- [ ] Bugfix
- [x] Enhancement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause the existing functionality to not work as expected)

## How Has This Been Tested?

- [ ] New unit tests added.
- [ ] Manual tested.

## Checklist:

- [x] Using Branch Name Convention
    - `feature/JIRA-ID-SHORT-DESCRIPTION` if has a JIRA ticket
    - `enhancement/SHORT-DESCRIPTION` if has/has no JIRA ticket and contain enhancement
    - `hotfix/SHORT-DESCRIPTION` if the change doesn't need to be tested (urgent)
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have made the documentation for the corresponding changes.

### CODE REVIEW

File: src/akushon/action/node/action_manager.cpp [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+    bool is_load_success = false;
+    try {
+      std::ifstream file(path);
+      nlohmann::json action_data = nlohmann::json::parse(file);
+
+      action.set_name(action_data["name"]);
+
+      for (auto [key, val] : action_data.items()) {
+        if (key.find("step_") != std::string::npos) {
```
Comment (segara2410) [inline src/akushon/action/node/action_manager.cpp:79]: should be changed later, shouldn't strict the name of a pose with prefix `step_`

File: src/akushon/action/process/joint_process.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+JointProcess::JointProcess(uint8_t joint_id, float position)
+: joint(tachimawari::joint::Joint(joint_id, position)), initial_position(position),
+  target_position(position), additional_position(0.0)
+{
+}
+
+void JointProcess::set_target_position(float target_position, float speed)
+{
+  this->target_position = target_position;
+  additional_position = (this->target_position - initial_position) * speed;
```
Comment (segara2410) [inline src/akushon/action/process/joint_process.cpp:40]: limit the minimum speed to 0 and maximum speed to 1

File: src/akushon/action/node/action_node.cpp [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    auto joint_msg = tachimawari_interfaces::msg::Joint();
+
+    joint_msg.id = joint.get_id();
+    joint_msg.position = joint.get_position();
+
+    joints.push_back(joint_msg);
+  }
+
+  auto joints_msg = tachimawari_interfaces::msg::SetJoints();
+  joints_msg.joints = joints;
```
Comment (threeal) [inline src/akushon/action/node/action_node.cpp:143]: if performance is important, we could do this instead
```c++
auto joints_msg = tachimawari_interfaces::msg::SetJoints();

const auto & joints = action_manager->get_joints();
auto & joint_msgs = joints_msg.joints;

joint_msgs.resize(joints.size());
for (size_t i = 0; i < joints.size() && i < joints_msg.size(); ++i) {
  joint_msgs[i].id = joints[i].get_id();
  // ...
}
```

File: src/akushon/action/process/interpolator.cpp [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+Interpolator::Interpolator(const std::vector<Action> & actions, const Pose & initial_pose)
+: actions(actions), joint_processes({}), current_pose_index(0),
+  pause_time(0), init_pause(false), start_stop_time(0), init_state(true),
+  current_action_index(0)
+{
+  for (const auto & joint : initial_pose.get_joints()) {
+    joint_processes.insert(
+      {joint.get_id(),
+        JointProcess(joint.get_id(), joint.get_position())});
```
Comment (threeal) [inline src/akushon/action/process/interpolator.cpp:41]: maybe use this to simplify lines?
```c++
const auto & joint_process = JointProcess(joint.get_id(), joint.get_position());
joint_processes.insert({joint.get_id(), joint_process});
```

File: src/akushon/action/process/interpolator.cpp [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+          if (init_pause) {
+            init_pause = false;
+            pause_time = time;
+          }
+
+          if (current_pose_index == actions[current_action_index].get_pose_count()) {
+            change_state(STOP_DELAY);
+            init_pause = true;
+          } else if ((time - pause_time) >  // NOLINT
+            (actions[current_action_index].get_pose(current_pose_index).get_pause() * 1000))
```
Comment (threeal) [inline src/akushon/action/process/interpolator.cpp:80]: can we save `actions[current_action_index]` to shorten lines? (and prevent linter error)
```c++
const auto & current_action = actions[current_action_index];
if (current_pose_index == current_action .get_pose_count()) {
  // ...
} else if ((time - pause_time) >  (current_action .get_pose(current_pose_index).get_pause() * 1000)) {
  // ...
}
```
or save it even more
```c++
const auto & current_action = actions[current_action_index];
if (current_pose_index == current_action .get_pose_count()) {
  // ...
} else {
  const auto & current_pose = current_action .get_pose(current_pose_index)
  if ((time - pause_time) >  (current_pose.get_pause() * 1000)) {
    // ...
  }
}
```

File: src/akushon/action/process/interpolator.cpp [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+    case STOP_DELAY:
+      {
+        if (init_state) {
+          init_state = false;
+          start_stop_time = time;
+        }
+
+        if ((time - start_stop_time) > (actions[current_action_index].get_stop_delay() * 1000)) {
+          current_action_index++;
```
Comment (threeal) [inline src/akushon/action/process/interpolator.cpp:98]: can we use `++current_action_index` instead?

see these lines too
https://github.com/ichiro-its/akushon/blob/38c8be3306632df9ed4a87c0c913b599c676608c/src/akushon/action/process/interpolator.cpp#L130
https://github.com/ichiro-its/akushon/blob/38c8be3306632df9ed4a87c0c913b599c676608c/src/akushon/action/process/interpolator.cpp#L138

File: src/akushon/action/process/interpolator.cpp [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+bool Interpolator::is_finished() const
+{
+  return state == END;
+}
+
+void Interpolator::next_pose()
+{
+  for (const auto & joint :
+    actions[current_action_index].get_pose(current_pose_index).get_joints())
```
Comment (threeal) [inline src/akushon/action/process/interpolator.cpp:124]: can we save `actions[current_action_index]` to shorten content of `for` ?
```c++
const auto & current_action = actions[current_action_index];
for (const auto & joint : current_action.get_pose(current_pose_index).get_joints()) {
  // ...
}
```

or better, create a function from it (if it always get called)
```c++
const auto & get_current_action() const {
  return actions[current_action_index];
}

const auto & get_current_pose() const {
  return get_current_action().get_pose(current_pose_index);
}
```

File: src/akushon/node/akushon_node.cpp [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        const std::shared_ptr<GoalHandleRunAction> goal_handle) -> rclcpp_action::CancelResponse {
+        return rclcpp_action::CancelResponse::ACCEPT;
+      },
+      [this](const std::shared_ptr<GoalHandleRunAction> goal_handle) {
+        std::thread{[this](const std::shared_ptr<GoalHandleRunAction> goal_handle) {
+            rclcpp::Rate rcl_rate(8ms);
+
+            const auto goal = goal_handle->get_goal();
+            auto feedback = std::make_shared<RunAction::Feedback>();
+            auto result = std::make_shared<RunAction::Result>();
```
Comment (threeal) [inline src/akushon/node/akushon_node.cpp:79]: can we put this close to it's usages

File: src/akushon/node/akushon_node.cpp [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+                goal_handle->publish_feedback(feedback);
+              }
+            }
+
+            if (rclcpp::ok()) {
+              if (is_ready) {
+                result->status = SUCCEEDED;
+              } else {
+                result->status = FAILED;
+              }
```
Comment (threeal) [inline src/akushon/node/akushon_node.cpp:113]: could be simplified to `result->status = is_ready ? SUCCEEDED : FAILED;`

File: src/akushon/node/akushon_node.cpp [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+              goal_handle->succeed(result);
+            }
+          }, goal_handle}.join();
+      });
```
Comment (threeal) [inline src/akushon/node/akushon_node.cpp:117]: can we put this lambdas as functions instead? looks like it has to many lambdas to be handled.

```c++
using namespace std::placeholders;
run_action_server = rclcpp_action::create_server<RunAction>(
  node->get_node_base_interface(),
  node->get_node_clock_interface(),
  node->get_node_logging_interface(),
  node->get_node_waitables_interface(),
  "run_action",
  std::bind(this, &AkushonNode::first_function, _1, _2),
  std::bind(this, &AkushonNode::second_function, _1),
  /*...*/);
```

File: src/akushon/node/akushon_node.cpp [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+          } else if (goal->action_name != "") {
+            is_action_exist = action_node->is_action_exist(goal->action_name);
+          }
+        }
+
+        if (is_action_exist) {
+          return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
+        } else {
+          return rclcpp_action::GoalResponse::REJECT;
+        }
```
Comment (threeal) [inline src/akushon/node/akushon_node.cpp:67]: (optional) should we write it like this?
```c++
return is_action_exist
  ? rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE
  : rclcpp_action::GoalResponse::REJECT;
```

**Review by threeal — CHANGES_REQUESTED** [review]
some comments are optional though...

File: src/akushon/action/process/joint_process.cpp (hunk 2 above)
Comment (maroqijalil) [reply src/akushon/action/process/joint_process.cpp]: done

File: .gitignore [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+/planning/docs
+/planning/src
+
+*~
+
+# Emacs
+.#*
+
+# Catkin custom files
+CATKIN_IGNORE
```
[the remark this answers is not in the document]
Comment (maroqijalil) [reply .gitignore]: Maybe, I choose to remove it

File: src/akushon/action/node/action_manager.cpp [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+#include <vector>
+
+#include "akushon/action/model/action_name.hpp"
+#include "akushon/action/node/action_manager.hpp"
+#include "akushon/action/process/interpolator.hpp"
+#include "nlohmann/json.hpp"
+#include "tachimawari/joint/joint.hpp"
+
+#include "stdlib.h"  // NOLINT
+#include "unistd.h"  // NOLINT
```
[the remark this answers is not in the document]
Comment (maroqijalil) [reply src/akushon/action/node/action_manager.cpp]: I remove it

File: src/akushon/action/node/action_manager.cpp [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+const Action & ActionManager::get_action(int action_id) const
+{
+  if (actions.find(action_id) != actions.end()) {
+    return actions.at(action_id);
+  }
+
+  return Action("");
+}
```
[the remark this answers is not in the document]
Comment (maroqijalil) [reply src/akushon/action/node/action_manager.cpp]: done

File: src/akushon/action/node/action_manager.cpp [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+
+      for (auto [key, val] : action_data.items()) {
+        if (key.find("step_") != std::string::npos) {
+          Pose pose(key);
+          std::vector<tachimawari::joint::Joint> joints;
+
+          for (auto [steps_key, steps_val] : action_data[key].items()) {
+            if (!(steps_key.find("step_") != std::string::npos)) {
+              tachimawari::joint::Joint joint(
+                tachimawari::joint::JointId::by_name.at(steps_key),
```
[the remark this answers is not in the document]
Comment (maroqijalil) [reply src/akushon/action/node/action_manager.cpp]: I add some condition to handle it

File: src/akushon/action/node/action_node.cpp (hunk 3 above)
Comment (maroqijalil) [reply src/akushon/action/node/action_node.cpp]: done

File: src/akushon/action/process/interpolator.cpp (hunk 7 above)
Comment (maroqijalil) [reply src/akushon/action/process/interpolator.cpp]: done

File: src/akushon/node/akushon_node.cpp (hunk 10 above)
Comment (maroqijalil) [reply src/akushon/node/akushon_node.cpp]: done

**Comment by maroqijalil** [comment]
I've done all the suggestions and requests for changes, I did some optional changes as well, although I didn't put the reply for all the comments. Re-review this if there are other suggestions or requests for changes.

File: src/akushon/action/model/action.cpp [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+
+int Action::get_stop_delay() const
+{
+  return stop_delay;
+}
+
+void Action::set_next_action(std::string next_action)
+{
+  this->next_action = next_action;
+}
```
Comment (threeal) [inline src/akushon/action/model/action.cpp:101]: couldn't we pass `next_action` as `const std::string &` instead?

File: src/akushon/action/process/interpolator.cpp [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+    case PLAYING:
+      {
+        if (check_for_next()) {
+          if (init_pause) {
+            init_pause = false;
+            pause_time = time;
+          }
+
+          const auto & current_action = get_current_action();
+          if (current_pose_index == current_action.get_pose_count()) {
```
Comment (threeal) [inline src/akushon/action/process/interpolator.cpp:76]: (optional) seems it's fine to just call `get_current_action().get_pose_count()`.

File: src/akushon/node/akushon_node.cpp [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+
+      bool is_ready = false;
+      if (goal->action_id >= 0) {
+        is_ready = action_node->start(goal->action_id);
+      } else if (goal->action_name != "") {
+        is_ready = action_node->start(goal->action_name);
+      }
+
+      auto feedback = std::make_shared<RunAction::Feedback>();
+      auto result = std::make_shared<RunAction::Result>();
```
Comment (threeal) [inline src/akushon/node/akushon_node.cpp:96]: can we put this close to it's usages

File: src/akushon/node/akushon_node.cpp [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+#include "rclcpp_action/rclcpp_action.hpp"
+
+using namespace std::chrono_literals;
+using namespace std::placeholders;
+
+namespace akushon
+{
+
+using akushon_interfaces::action::RunAction;
+using GoalHandleRunAction = rclcpp_action::ServerGoalHandle<RunAction>;
```
Comment (threeal) [inline src/akushon/node/akushon_node.cpp:40]: this one already declared here:
https://github.com/ichiro-its/akushon/blob/c1efaad1bd4038f24fa3945ea51da280dc7630fb/include/akushon/node/akushon_node.hpp#L36-L37

you could use that instead, but inside `akushon` namespace

**Review by threeal — APPROVED** [review]
looks good though 👍

**Review by segara2410 — APPROVED** [review]

**Pull request merged**